### PR TITLE
[SPARKR-150] phase 1: implement sortBy() and sortByKey().

### PR DESCRIPTION
**Pull request by sun-rui**
This is phase 1 patch for SPARKR-150. It implements sortBy() and sortByKey().

Basically the rangePartitioner is the same as that in PySpark. It is simpler that the rangePartitioner in Spark, we can enhance it later if necessary.

Now add only two test cases. Will add more test cases during review.

### CODE REVIEW

File: pkg/R/RDD.R [hunk 1]
```diff
@@ -1240,6 +1240,40 @@ setMethod("flatMapValues",
...
+#' collect(sortBy(rdd, function(x) { x })) # list (1, 2, 3)
+#'}
+setGeneric("sortBy", function(rdd, func, ascending, numPartitions) { standardGeneric("sortBy") })
+
+setClassUnion("missingOrLogical", c("missing", "logical"))
+#' @rdname sortBy
+#' @aliases sortBy,RDD,RDD-method
+setMethod("sortBy",
+          signature(rdd = "RDD", func = "function", 
+                    ascending = "missingOrLogical", numPartitions = "missingOrInteger"),
```
Comment (shivaram) [inline pkg/R/RDD.R]: Can't we just use default argument for `ascending` ? (Something like in `collect(..., flatten=TRUE)`). As far as I remember we needed the `missingOrInteger` because the default argument depended on the RDD for `numPartitions`.

File: pkg/R/RDD.R [hunk 2]
```diff
@@ -1796,6 +1830,78 @@ setMethod("cogroup",
...
+#' sc <- sparkR.init()
+#' rdd <- parallelize(sc, list(list(3, 3), list(2, 2), list(1, 1)))
+#' collect(sortByKey(rdd)) # list (list(1, 1), list(2, 2), list(3, 3))
+#'}
+setGeneric("sortByKey", function(rdd, ascending, numPartitions) { standardGeneric("sortByKey") })
+
+#' @rdname sortByKey
+#' @aliases sortByKey,RDD,RDD-method
+setMethod("sortByKey",
+          signature(rdd = "RDD", ascending = "missingOrLogical", numPartitions = "missingOrInteger"),
```
Comment (shivaram) [inline pkg/R/RDD.R]: I know that `ascending` is in line with the RDD API, but I am wondering if `decreasing = FALSE` is better here as its more similar to R's `sort` function. Any thoughts ?

File: pkg/R/RDD.R [hunk 3]
```diff
@@ -1796,6 +1830,78 @@ setMethod("cogroup",
...
+                                      function(i) {
+                                        j <- ceiling(length(samples) * i / numPartitions)
+                                        samples[j]
+                                      })
+              }
+            }
+
+            rangePartitionFunc <- function(key) {
+              partition <- 0
+
```
Comment (shivaram) [inline pkg/R/RDD.R]: You can add a `TODO` here that we should try to use binary search instead of linear search. We can try to implement a utils function to do that ?

**Comment by shivaram** [comment]
Thanks @sun-rui for this change. This is looking pretty good ! I left some minor comments inline.

@concretevitamin -- it would be great if you could also take a look

File: pkg/R/RDD.R (hunk 1 above)
Comment (sun-rui) [inline pkg/R/RDD.R]: Yes, I did try using default arguments in setMethod. But seems there is bug or something I don't know that make the default arguments not work in setMethod. This actually is a workaround which seems a little bit ugly. I will do more investigation to see if I can use default arguments.

File: pkg/R/RDD.R (hunk 2 above)
Comment (sun-rui) [inline pkg/R/RDD.R]: I think consistency with RDD API is more important. It won't confusing programmers with RDD API in mind.

File: pkg/R/RDD.R (hunk 3 above)
Comment (sun-rui) [inline pkg/R/RDD.R]: OK. I will add a TODO

File: pkg/R/RDD.R (hunk 1 above)
Comment (shivaram) [inline pkg/R/RDD.R]: I think the trick is we can't dispatch with the optional argument. i.e. we can't have `ascending` in the `setGeneric` method. Also we might need to make it the last argument to get it to work

File: pkg/R/RDD.R [hunk 4]
```diff
@@ -1796,6 +1830,79 @@ setMethod("cogroup",
...
+#' @param rdd An (k, v) pair RDD to be sorted.
+#' @param ascending A flag to indicate whether the sorting is ascending or descending.
+#' @param numPartitions Number of partitions to create.
+#' @return An RDD where all (k, v) pair elements are sorted.
+#' @rdname sortByKey
+#' @export
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' rdd <- parallelize(sc, list(list(3, 3), list(2, 2), list(1, 1)))
```
Comment (concretevitamin) [inline pkg/R/RDD.R]: I think it'd be more educational to users if we, say, have key-val inversions (e.g. (3, 1), ...).

File: pkg/R/RDD.R (hunk 2 above)
Comment (concretevitamin) [inline pkg/R/RDD.R]: +1 on keeping the RDD API

File: pkg/R/RDD.R [hunk 5]
```diff
@@ -1796,6 +1830,79 @@ setMethod("cogroup",
...
+            
+            if (numPartitions > 1) {
+              rddSize <- count(rdd)
+              # constant from Spark's RangePartitioner
+              maxSampleSize <- numPartitions * 20
+              fraction <- min(maxSampleSize / max(rddSize, 1), 1.0)
+              
+              samples <- collect(keys(sampleRDD(rdd, FALSE, fraction, 1L)))
+              
+              # Note: the built-in R sort() function only atomic vectors
```
Comment (concretevitamin) [inline pkg/R/RDD.R]: missing "...works on..."?

File: pkg/R/RDD.R [hunk 6]
```diff
@@ -1796,6 +1830,79 @@ setMethod("cogroup",
...
+                                        j <- ceiling(length(samples) * i / numPartitions)
+                                        samples[j]
+                                      })
+              }
+            }
+
+            rangePartitionFunc <- function(key) {
+              partition <- 0
+              
+              // TODO: Use binary search instead of linear search, similar with Spark
```
Comment (concretevitamin) [inline pkg/R/RDD.R]: Use # instead of // ?

File: pkg/inst/tests/test_rdd.R [hunk 7]
```diff
@@ -254,6 +254,17 @@ test_that("keyBy on RDDs", {
   expect_equal(actual, lapply(nums, function(x) { list(func(x), x) }))
 })
 
+test_that("sortBy() on RDDs", {
```
Comment (concretevitamin) [inline pkg/inst/tests/test_rdd.R]: Would be good to test on a non-identity function.

File: pkg/inst/tests/test_rdd.R [hunk 8]
```diff
@@ -373,3 +384,17 @@ test_that("fullOuterJoin() on pairwise RDDs", {
...
+  actual <- collect(sortedRdd)
+  numPairs <- lapply(nums, function(x) { list (x, x) })
+  expect_equal(actual, sortKeyValueList(numPairs, decreasing = TRUE))
+
+  rdd2 <- parallelize(sc, sort(nums, decreasing = TRUE), 2L)
+  numPairsRdd2 <- map(rdd2, function(x) { list (x, x) })
+  sortedRdd2 <- sortByKey(numPairsRdd2)
+  actual <- collect(sortedRdd2)
+  expect_equal(actual, numPairs)
+})
```
Comment (concretevitamin) [inline pkg/inst/tests/test_rdd.R]: newline at EOF

File: pkg/inst/tests/test_rdd.R [hunk 9]
```diff
@@ -373,3 +384,17 @@ test_that("fullOuterJoin() on pairwise RDDs", {
   expect_equal(sortKeyValueList(actual),
                sortKeyValueList(list(list("a", list(1, NULL)), list("b", list(2, NULL)), list("d", list(NULL, 4)), list("c", list(NULL, 3)))))
 })
+
+test_that("sortByKey() on pairwise RDDs", {
+  numPairsRdd <- map(rdd, function(x) { list (x, x) })
```
Comment (concretevitamin) [inline pkg/inst/tests/test_rdd.R]: Would be good to have string keys?

File: pkg/R/RDD.R (hunk 3 above)
Comment (concretevitamin) [inline pkg/R/RDD.R]: Does R have some binary search built in that we can use as an one-liner? If it does then it'd be good to add it.

**Comment by concretevitamin** [comment]
I did a pass and the patch looks good - thanks @sun-rui. A comment on testing: it'd be good to test some boundary cases where there's only one element and/or one partition in an RDD, etc.

File: pkg/R/RDD.R (hunk 4 above)
Comment (sun-rui) [inline pkg/R/RDD.R]: fixed

File: pkg/R/RDD.R (hunk 5 above)
Comment (sun-rui) [inline pkg/R/RDD.R]: fixed

File: pkg/R/RDD.R (hunk 6 above)
Comment (sun-rui) [inline pkg/R/RDD.R]: fixed

File: pkg/inst/tests/test_rdd.R (hunk 7 above)
Comment (sun-rui) [inline pkg/inst/tests/test_rdd.R]: fixed. Use function (x) { x * x }.

File: pkg/inst/tests/test_rdd.R (hunk 8 above)
Comment (sun-rui) [inline pkg/inst/tests/test_rdd.R]: fixed

File: pkg/inst/tests/test_rdd.R (hunk 9 above)
Comment (sun-rui) [inline pkg/inst/tests/test_rdd.R]: fixed

**Comment by sun-rui** [comment]
thanks @concretevitamin. Add more tests on boundary cases.

File: pkg/R/RDD.R (hunk 1 above)
Comment (sun-rui) [inline pkg/R/RDD.R]: Use default arguments for sortBy() and sortKeyBy().

File: pkg/R/RDD.R (hunk 3 above)
Comment (sun-rui) [inline pkg/R/RDD.R]: R does not have built-in binary search function. While there are two third-party packages, like gtools and data.table, providing binary search functions, they are GPL licensed.

**Comment by concretevitamin** [comment]
@sun-rui could you rebase?

**Comment by sun-rui** [comment]
Merged with upstream master.

**Comment by shivaram** [comment]
Thanks @sun-rui - LGTM. @concretevitamin - Any further comments ?

**Comment by concretevitamin** [comment]
LGTM - merging this.
